python/BioSimSpace/Parameters/_parameters.py

```diff
@@ -144,19 +155,25 @@ def ff99(molecule, water_model=None, leap_commands=None, work_dir=None, property
                                     "GROMACS (http://www.gromacs.org).")
 
     # Validate arguments.
-    _validate(molecule=molecule, water_model=water_model,
-        check_ions=True, leap_commands=leap_commands, property_map=property_map)
+    _validate(molecule=molecule, tolerance=tolerance, max_distance=max_distance,
+        water_model=water_model, check_ions=True, leap_commands=leap_commands,
+        bonds=bonds, property_map=property_map)
 
     # Create a default protocol.
-    protocol = _Protocol.FF99(water_model=water_model,
+    protocol = _Protocol.FF99(tolerance=tolerance,
+                              water_model=water_model,
+                              max_distance=max_distance,
                               leap_commands=leap_commands,
+                              bonds=bonds,
                               property_map=property_map)
 
     # Run the parameterisation protocol in the background and return
     # a handle to the thread.
     return _Process(molecule, protocol, work_dir=work_dir, auto_start=True)
 
-def ff99SB(molecule, water_model=None, leap_commands=None, work_dir=None, property_map={}):
+def ff99SB(molecule, tolerance=1.2, max_distance=_Length(6, "A"),
```

Comment:
There's a lot of similarity between at least 5 of the classes (`ff99`, `ff99SB`, `ff99SBildn`, `ff03` and `ff14SB`. Seeing that the only difference between them is basically the import statement in `tleap`, I think it might make sense to have a base `factory` class or something along these lines, that you just subclass in a trivial manner, so that the `ff99` class only calls `FF99` with some parameter (or even better, abstracted FF class with a parameter of `"protein_force_field"). Would make maintenance much easier because it would remove a lot of duplicate code (that was my impression anyway, correct me if the duplication is needed).

Reply:
These aren't classes, rather functions that call through to a generic AMBER `tLEaP` protocol that does essentially what you say, i.e. searches for the force field name in the `dat` directory (using some logic to search for deprecated force fields) and adds it to the `LEaP` script if found. (Perhaps it's the protocols that you are referring to.)

When I originally wrote this I had never parameterised a molecule before so wasn't sure how different the protocols would be for the different force fields, e.g. what options would be supported. I also wasn't sure what would and wouldn't be supported by `pdb2gmx`, which is only used as a fallback. This is why there are separate protocols for each named force field, although they all (at least the protein ones) use the same base `run` method to execute the protocol.

I think it is the _protocols_ that would likely be better abstracted in the way that you suggest, i.e. having protocols for protein and ligand / small molecule force fields, with ones for the different engines that are supported.

I'll have a go a refactoring things since I agree that this part of the code is quite messy as is.